lib/Sema/LookupVisibleDecls.cpp

```diff
@@ -478,6 +474,56 @@ static void
   lookupTypeMembers(BaseTy, PT, Consumer, CurrDC, LS, Reason);
 }
 
+/// Trigger synthesizing implicit member declarations to make them "visible".
+static void synthesizeMemberDeclsForLookup(NominalTypeDecl *NTD,
+                                           const DeclContext *DC) {
+  // Synthesize the memberwise initializer for structs or default initializer
+  // for classes.
+  if (!NTD->hasInterfaceType())
```

Comment:
It's possible that an interface type has been set, but implicit constructors were not added (and vice versa). You can unconditinally call `addImplicitConstructors()` here; it's idempotent if called multiple times.

Reply:
IIRC, that causes circular dependency error. I will try that again.

Reply:
Here's the cyclic dependency error when enabling typo correction. How can we fix this?

```
$ cat test.swift
struct Broken {
  var b = True
}

$ swift -frontend -typecheck -typo-correction-limit 10 -debug-cycles test.swift
test.swift:1:8: error: circular reference
struct Broken {
       ^
test.swift:2:3: note: through reference here
  var b = True
  ^
test.swift:2:7: note: through reference here
  var b = True
      ^
test.swift:2:7: note: through reference here
  var b = True
      ^
===CYCLE DETECTED===
 `--SynthesizeMemberwiseInitRequest(test.(file).Broken@test.swift:1:8)
     `--ParseMembersRequest(test.(file).Broken@test.swift:1:8) -> {(unknown decl), test.(file).Broken.b@test.swift:2:7}
     `--StorageImplInfoRequest(test.(file).Broken.b@test.swift:2:7)
     |   `--AttachedPropertyWrappersRequest(test.(file).Broken.b@test.swift:2:7)
     `--StorageImplInfoRequest(test.(file).Broken.b@test.swift:2:7)
     |   `--AttachedPropertyWrappersRequest(test.(file).Broken.b@test.swift:2:7)
     `--AccessLevelRequest(test.(file).Broken.b@test.swift:2:7)
     `--AccessLevelRequest(test.(file).Broken.b@test.swift:2:7)
     `--InterfaceTypeRequest(test.(file).Broken.b@test.swift:2:7)
         `--NamingPatternRequest(test.(file).Broken.b@test.swift:2:7)
             `--PatternBindingEntryRequest((unknown decl), 0)
                 `--AttachedPropertyWrappersRequest(test.(file).Broken.b@test.swift:2:7)
 ...
                 `--ParseMembersRequest(test.(file).Broken@test.swift:1:8) -> {(unknown decl), test.(file).Broken.b@test.swift:2:7}
                 `--GenericParamListRequest(test.(file).Broken@test.swift:1:8)
                 `--GenericSignatureRequest(test.(file).Broken@test.swift:1:8)
                 |   `--GenericParamListRequest(test.(file).Broken@test.swift:1:8)
                 `--GenericParamListRequest(test.(file).Broken@test.swift:1:8)
                 `--InterfaceTypeRequest(test.(file).Broken@test.swift:1:8)
                 |   `--GenericParamListRequest(test.(file).Broken@test.swift:1:8)
                 |   `--GenericParamListRequest(test.(file).Broken@test.swift:1:8)
                 `--HasMemberwiseInitRequest(test.(file).Broken@test.swift:1:8) -> 1
                 |   `--InterfaceTypeRequest(test.(file).Broken@test.swift:1:8)
                 |   |   `--GenericParamListRequest(test.(file).Broken@test.swift:1:8)
                 |   |   `--GenericParamListRequest(test.(file).Broken@test.swift:1:8)
                 |   `--HasUserDefinedDesignatedInitRequest(test.(file).Broken@test.swift:1:8) -> 0
                 |   |   `--ParseMembersRequest(test.(file).Broken@test.swift:1:8) -> {(unknown decl), test.(file).Broken.b@test.swift:2:7}
                 |   `--ParseMembersRequest(test.(file).Broken@test.swift:1:8) -> {(unknown decl), test.(file).Broken.b@test.swift:2:7}
                 |   `--StorageImplInfoRequest(test.(file).Broken.b@test.swift:2:7)
                 |   |   `--AttachedPropertyWrappersRequest(test.(file).Broken.b@test.swift:2:7)
                 |   `--StorageImplInfoRequest(test.(file).Broken.b@test.swift:2:7)
                 |   |   `--AttachedPropertyWrappersRequest(test.(file).Broken.b@test.swift:2:7)
                 |   `--AccessLevelRequest(test.(file).Broken.b@test.swift:2:7)
                 `--SynthesizeMemberwiseInitRequest(test.(file).Broken@test.swift:1:8) (cyclic dependency)
```

Reply:
I see. It might be better then to check to explicitly check if an SynthesizeMemberwiseInitRequest is active, instead of using `hasInterfaceType()`, ie,

```
VD->getASTContext().evaluator.hasActiveRequest(
            SynthesizeMemberwiseInitRequest{...})
```